### DISCUSSION TITLE
How do I trigger a source generator from a *.razor file?

### DISCUSSION THREAD

**Discussion by MulletPr00f** [dotnet/roslyn#75048] [Q&A] [answered]
As the title says, I'm trying to figure out how to trigger a call to my ISourceGenerator implementation when I modify and save a *.razor file. As of now, I've only managed to trigger it when I modify and save a *.cs file.

**Comment by 333fred** [comment]
Well, first and foremost, you shouldn't be using ISourceGenerator. You should be using IIncrementalGenerator, see https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md and https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.cookbook.md. In an incremental generator, you'd base your generator on the additional files provider, and go from there.

**Reply by MulletPr00f** [reply]
Hi, thanks for the reply.

I've modified my code to use IIncrementalGenerator as you pointed, using the same example in the cookbook (I've only changed the content of the generated file to a commented `DateTime.Now`):

```using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Text;
using System;
using System.IO;
using System.Text;

namespace MyGenerators
{
    [Generator]
    public class BlazorDesignerIncrementalGenerator : IIncrementalGenerator
    {
        public void Initialize(IncrementalGeneratorInitializationContext context)
        {
            var pipeline = context.AdditionalTextsProvider
                .Where(static (text) => text.Path.EndsWith(".razor"))
                .Select(static (text, cancellationToken) =>
                {
                    var name = Path.GetFileName(text.Path);
                    var content = $"// {DateTime.Now.ToString("HH:mm:ss")}";
                    return (name, content);
                });

            context.RegisterSourceOutput(pipeline,
                static (context, pair) =>
                    // Note: this AddSource is simplified. You will likely want to include the path in the name of the file to avoid
                    // issues with duplicate file names in different paths in the same project.
                    context.AddSource($"{pair.name}generated.cs", SourceText.From(pair.content, Encoding.UTF8)));
        }
    }
}
```

Unfortunately my issue continues. `Initialize` is only being called when I modify and save a *.cs file. If I modify a *.razor file, it doesn't trigger my BlazorDesignerIncrementalGenerator.

I want to be able to Ctrl+S a *.razor file and it calls my BlazorDesignerIncrementalGenerator. Is it possible? What an I missing?

**Reply by 333fred** [reply]
Do you have your VS set to run generators in balanced mode?

**Reply by 333fred** [reply]
If not, perhaps @jasonmalinowski will have a better idea.

**Reply by MulletPr00f** [reply]
Yes, but I've tried both (relaunching VS):

![image](https://github.com/user-attachments/assets/f7c7b8f3-c121-47b8-b4d3-740c54a03573)

This config shouldn't apply only for C# editor?

**Reply by 333fred** [reply]
It applies to everything, afaik, but we'll have to wait for @jasonmalinowski or @CyrusNajmabadi to chime in.

**Reply by MulletPr00f** [reply]
Any way to configure my generator to run when I save a *.razor file, @jasonmalinowski @CyrusNajmabadi ?

**Reply by jasonmalinowski** [reply]
> Unfortunately my issue continues. Initialize is only being called when I modify and save a *.cs file. If I modify a *.razor file, it doesn't trigger my BlazorDesignerIncrementalGenerator.

@MulletPr00f Did you mean the initialize method is being called once, or your lambda in the Select call is only being called once?

**Reply by jasonmalinowski** [reply]
If you're in automatic and it's still only running on save, then that'd make me think we have some bug where we're not wiring up the additional file to the editor. But it seems like we'd have a lot more broken if that was the case...

**Reply by MulletPr00f** [reply]
@jasonmalinowski it's not running on save. Only if I save a *.cs file. If I save a *.razor file, it's not running. 

The *.razor files are being added as additional files by the Blazor SDK. If I add *.txt files as additional files, both file types are listed as additional files in the generators, but they don't trigger the generators on file save.

**Reply by MulletPr00f** [reply]
Turns out that I need to get out of the car and get in again for this to work. Just as I did previously when changing balenced mode to automatic, I had to relaunch VS after a build on my generator's library and remove the reference to it in my client project and readd it. Then it starting triggering the lambda expression on *.razor files, **but only in automatic.** On balanced mode I have still no luck. Restaring it again didn't help either.

**Reply by 333fred** [reply]
So, the balanced-mode having this behavior, while a bug, doesn't honestly _surprise_ me. @MulletPr00f do you want to open an issue for that?

**Reply by MulletPr00f** [reply]
@333fred what concerns me is that this option affects _all_ generators, not only mine. I'm writing this generator to a team of 150+ developers, so if it starts to slow down their coding environments, I'm gonna hear it. On top of that, it wouldn't be very nice to remind all of the them to change this setting.

So yes, I'd like open an issue for future investigation, please.

**Reply by 333fred** [reply]
> so if it starts to slow down their coding environments, I'm gonna hear it

If it does, you can encourage them to file feedback. However, we've generally noticed that people don't actually care that their source generators are a tiny bit out of date until a file is actually saved, and by reducing the amount of working happening in the background, we significantly improve the rest of the experience of using VS by simply making _that_ faster. The lack of saving a razor file triggering a rebuild in balanced mode is a bug, so feel free to [open one](https://github.com/dotnet/roslyn/issues/new/choose) for that.

**Reply by jasonmalinowski** [reply]
So saving additional files should be triggering generators in even balanced mode:

https://github.com/dotnet/roslyn/blob/7b7951aa13c50ad768538e58ed3805898b058928/src/VisualStudio/Core/Def/ProjectSystem/VisualStudioWorkspaceImpl_SourceGenerators.cs#L68-L70

If we're not seeing that, then yes please do file a bug.

**Reply by CyrusNajmabadi** [reply, marked as answer]
@jasonmalinowski this was a recent bug fix to support saving additional files.  Previously, we only hooked up to actual Roslyn docs (c#/vb etc)

**Reply by 333fred** [reply]
Looks like this was added to 17.12. @MulletPr00f, are you using 17.11? If so, then this is already fixed when you upgrade to 17.12 later this year.

**Reply by jasonmalinowski** [reply]
And it looks like the changes should already be in 17.12 Preview 1; @MulletPr00f if you want to try that that'll verify we're already fixed.

**Reply by MulletPr00f** [reply]
@CyrusNajmabadi Great!

@333fred Yes, I'm using 17.11.4.

@jasonmalinowski Alright, I'm gonna try that, and I'll get back to you to tell if the fix worked.

Thanks.

**Reply by MulletPr00f** [reply]
> And it looks like the changes should already be in 17.12 Preview 1; @MulletPr00f if you want to try that that'll verify we're already fixed.

Confirmed it's working in 17.12.0 Preview 2.0! Can I mark all your three answers as the discussion answers?

**Reply by 333fred** [reply]
Unfortunately github discussions only allow 1 answer.
